docs/architecture/adr-101-data-companion-pull-api.md

```diff
@@ -0,0 +1,370 @@
+    // committed yet.
+    int64 height = 1;
+}
+
+message GetBlockRequest {
+    // The height of the block to get. Set to 0 to return the latest block.
+    int64 height = 1;
+}
+
+message GetBlockResponse {
+    // Block data for the requested height.
+    tendermint.types.Block block = 1;
+}
+
+message GetBlockResultsRequest {
+    // The height of the block results to get. Set to 0 to return the latest
+    // block.
+    int64 height = 1;
+}
+
+message GetBlockResultsResponse {
+    // The height of the block whose results are provided in the remaining
+    // fields.
+    int64 height = 1;
+
+    // All events produced by the ABCI FinalizeBlock call.
+    repeated tendermint.abci.Event events = 2;
+
+    // All transaction results (and their associated events) produced by block
+    // execution.
+    repeated tendermint.abci.ExecTxResult tx_results = 3;
+
+    // Validator updates during block execution.
+    repeated tendermint.abci.ValidatorUpdate validator_updates = 4;
+
+    // Consensus parameter updates during block execution.
+    tendermint.types.ConsensusParams consensus_param_updates = 5;
+
+    // The app hash returned by the application after execution of all of the
+    // transactions in the block.
+    bytes app_hash = 6;
+}
+```
+
+There are several reasons as to why there are two separate gRPC calls for this
+data as opposed to just one:
+
+1. The quantity of data stored by each is application-dependent, and coalescing
+   the two types of data could impose significant overhead in some cases.
+2. The existing JSON-RPC API distinguishes between endpoints providing these two
+   types of data (`/block` and `/block_results`), and the `BlockService` should
+   simply provide a gRPC alternative to these two endpoints.
+3. Eventually, when we no longer need to store block results at all, we can
+   simply deprecate the `GetBlockResults` API call without affecting clients who
+   rely on `GetBlock`.
+
+#### Data Companion Service
+
+This gRPC service is the only novel service proposed for the data companion API,
+and effectively gives the data companion a say in how the node prunes its data.
+
+```protobuf
+syntax = "proto3";
+
+package tendermint.data_companion_service.v1;
+
+// DataCompanionService provides privileged access to specialized pruning
+// functionality on the CometBFT node to help optimize node storage.
+service DataCompanionService {
+    // SetRetainHeight notifies the node of the minimum height whose data must
+    // be retained by the node. This data includes block data and block
+    // execution results.
+    //
+    // Setting a retain height lower than a previous setting will result in an
+    // error.
+    //
+    // The lower of this retain height and that set by the application in its
+    // Commit response will be used by the node to determine which heights' data
+    // can be pruned.
+    rpc SetRetainHeight(SetRetainHeightRequest) returns (SetRetainHeightResponse) {}
```

Comment:
Since we are storing both block data and block results, I am wondering whether we should have a separate retain_height for block data and for abci result data? Also there was a suggestion to store the results separately from the block, it might not be a bad idea.